### PR TITLE
WIP feat(CrazyTestCase): Provide for PHPUnit 6 and add param assertions

### PR DESCRIPTION
**Pull request by dacgray**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,31 +1,25 @@
...
+
+cache:
+  directories:
+  - $HOME/.composer/cache
+
+language: php
+
+matrix:
+  include:
+  - php: 7.2
```
Comment (Solomonic) [inline .travis.yml]: Please add
`- php: nightly`
and

```
allow_failures:
- php: nightly

fast_finish: true
```

File: .travis.yml [hunk 2]
```diff
@@ -1,31 +1,25 @@
...
-  - travis_retry composer update --prefer-dist
-  - if [ -n "$PROCESS_CODECLIMATE" ];then echo 'zend_extension=xdebug.so' >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini; fi
-
-script: ./test/bin/ci.bash summary
-
-cache:
-  directories:
-    - $HOME/.composer/cache
\ No newline at end of file
+sudo: required
```
Comment (Solomonic) [inline .travis.yml]: Just to confirm: do we need sudo rights to make it work?

File: .travis.yml (hunk 1 above)
Comment (dacgray) [reply .travis.yml]: Added

File: .travis.yml [hunk 3]
```diff
@@ -1,31 +1,29 @@
...
+
+  allow_failures:
+  - php: nightly
+
+install:
+  - chmod +x ./scripts/travis_install.sh && ./scripts/travis_install.sh
+
+before_script:
+  - travis_retry composer self-update
+  - travis_retry composer update --prefer-dist
```
Comment (wmathes) [inline .travis.yml]: don't run update, trust composer.lock and just run composer install

File: src/CrazyTestCase.php [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+     * @param $fn
+     * @param $exceptionClass
+     * @param string|null $message
+     */
+    protected function assertExceptionThrown($fn, $exceptionClass, ?string $message = null)
+    {
+        try {
+            $fn();
+        }
+        catch (\Throwable $t) {
```
Comment (Solomonic) [inline src/CrazyTestCase.php]: you want to do something here

File: composer.json [hunk 5]
```diff
@@ -1,43 +1,45 @@
...
-		"squizlabs/php_codesniffer": "~2.5"
-	},
-
-	"scripts": {
-		"test": "vendor/bin/phpunit -c phpunit.dist.xml && echo \"\\n>> Checking code standards\\n\" && test/bin/phpcs.bash emacs",
-		"csfix": "test/bin/phpcbf.bash",
-		"cscheck": "test/bin/phpcs.bash emacs"
-	}
+    "name": "crazyfactory/phpunit-extensions",
+    "minimum-stability": "dev",
```
Comment (wmathes) [inline composer.json]: why `dev`? Isn't `"prefer-stable": true` enough?

this flag is only effective when run in the root project. So it will basically only affect your dev dependencies and fail for all others as the flag gets overwritten within depending packages. Is that what you want? :)

File: README.md [hunk 6]
```diff
@@ -1,2 +1,24 @@
...
+- assertParamIsNotOptional
+
+# Use in your Project
+
+```json
+{
+    "repositories": [
+        {
+            "type": "vcs",
+            "url": "https://github.com/crazyfactory/php-phpunit-extensions"
```
Comment (wmathes) [inline README.md]: nope, definitely not "use in your project" like this. It will come from gemfury and never ever from github directly :) Basically speaking: kick out the repositories part ^^

being public this could also be published to packagist, but there isn't really a benefit for that right now (and that should be a separate PR/issue anyway)

File: composer.json [hunk 7]
```diff
@@ -1,43 +1,45 @@
...
+        "psr-4": {
+            "CrazyFactory\\PHPUnitExtensions\\Tests\\": "tests/"
+        }
+    },
+    "require": {
+        "php": "^7.2"
+    },
+    "require-dev": {
+        "phpunit/phpunit": "^6.0",
+        "squizlabs/php_codesniffer": "3.2.3",
```
Comment (wmathes) [inline composer.json]: curious... is that required? i believe the sniffs-package will pull it in anyway

File: src/CrazyTestCase.php (hunk 4 above)
Comment (dacgray) [reply src/CrazyTestCase.php]: I just need $t.  I declare another variable as $t, but it's not necessary.

File: composer.json [hunk 8]
```diff
@@ -1,43 +1,45 @@
...
+    "scripts": {
+        "test": "phpunit -c phpunit.xml",
+        "lint": "phpcs --standard=phpcs.xml",
+        "lint:fix": "phpcbf --standard=phpcs.xml"
+    },
+    "config": {
+        "optimize-autoloader": true,
+        "platform": {
+            "php": "7.2"
+        }
```
Comment (wmathes) [inline composer.json]: similar issue as with minimum stability. this is only effective in the root package. While it may feel like adding a more specific configuration, it doesn't offer any larger benefits for us and just blows up the codebase. I would kick out the config block completely here as they're really only useful for a root package.

File: src/CrazyTestCase.php [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+namespace CrazyFactory\PHPUnitExtensions;
+
+use PHPUnit\Framework\Constraint\IsFalse;
+use PHPUnit\Framework\Constraint\IsInstanceOf;
+use PHPUnit\Framework\Constraint\IsTrue;
+use PHPUnit\Framework\TestCase;
+use ReflectionParameter;
+
+abstract class CrazyTestCase extends TestCase
```
[the remark this answers is not in the document]
Comment (dacgray) [reply src/CrazyTestCase.php]: You don't need abstract methods in an abstract class.  It just means it must be extedned from.  

Answer for java, but i think it's relevant: https://www.quora.com/What-is-the-use-of-abstract-class-with-zero-abstract-methods-in-java

File: phpcs.xml [hunk 10]
```diff
@@ -1,36 +1,10 @@
...
\ No newline at end of file
+<?xml version="1.0" encoding="UTF-8"?>
+<ruleset name="CrazyFactory Default Coding Standard">
+    <config name="installed_paths" value="vendor/crazyfactory/sniffs/src" />
+
+    <rule ref="CrazyFactory" />
+
+    <file>./src/</file>
+    <file>./tests</file>
+
```
Comment (wmathes) [inline phpcs.xml]: one ending in / and one doesn't? :)

File: src/CrazyTestCase.php (hunk 4 above)
Comment (Solomonic) [reply src/CrazyTestCase.php]: i see

File: phpunit.xml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+<phpunit bootstrap="vendor/autoload.php"
+         backupGlobals="false"
+         forceCoversAnnotation="true"
```
Comment (wmathes) [inline phpunit.xml]: do you really want this one or is it a CPF? :) 

basically meaning for every test case you're writing you'll have to specify which functions it covers and only in these code coverage is actually registered when running tests. This is a great feature, but it's also a huge pain especially when indirectly testing private methods by using public endpoints.

I guess we're not using it in this project at all, but please be aware that this is in! :) It will prick you hard when trying to set up codecov later on

File: src/CrazyTestCase.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (wmathes) [reply src/CrazyTestCase.php]: I think the intention here is to use this as a guard class which can't be used without inheriting from it again.

Sounds like valid use-case to me, even though strictly speaking you have a point. How about agreeing on that abstract methods from the parent class weren't implemented, thus still exposing abstract method which need to be overwritten in a subclass, and therefor validating the abstract-accessibility :)

File: tests/unit/CrazyTestCaseTest.php [hunk 12]
```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace CrazyFactory\MagicModels\Test;
```
Comment (wmathes) [inline tests/unit/CrazyTestCaseTest.php]: whoops :)

File: src/TestCaseHelper.php [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+{
+
+    /**
+     * @param string $class
+     * @param string $method
+     * @param int $paramPosition
+     * @return ReflectionParameter|null
+     * @throws \ReflectionException
+     */
+    public static function getParam(string $class, string $method, int $paramPosition): ?ReflectionParameter
```
Comment (wmathes) [inline src/TestCaseHelper.php]: could also just be part of CrazyTestCase as a private method. 

otherwise i'd say it's really just replacing a one-line with another one-liner...

File: README.md (hunk 6 above)
Comment (Solomonic) [reply README.md]: the github pull is rather for dev purpose the production.

File: .travis.yml (hunk 3 above)
Comment (dacgray) [reply .travis.yml]: ah... yep.

File: README.md (hunk 6 above)
Comment (dacgray) [reply README.md]: okay - @Solomonic can you help with creating the package?

File: src/CrazyTestCase.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (Solomonic) [reply src/CrazyTestCase.php]: I'm good with the protectionist approach, we don't need to include an additional abstract method

File: phpcs.xml (hunk 10 above)
Comment (dacgray) [reply phpcs.xml]: sharpest eyes in the east

File: composer.json (hunk 7 above)
Comment (dacgray) [reply composer.json]: Not required... BUT, crazyfactory/sniffs breaks with codesniffer 3.3.0.

image

Issue already added to sniffs.

Comment (dacgray) [reply composer.json]: travis also has this prob - so its environment agnostic.

File: README.md (hunk 6 above)
Comment (wmathes) [reply README.md]: @Solomonic when developing locally you should still not rely on the github link. You should use a [path-repository](https://getcomposer.org/doc/05-repositories.md#path) for that. It will basically symlink whatever directory you desire (like another local git repo) into your project.

File: phpunit.xml (hunk 11 above)
Comment (dacgray) [reply phpunit.xml]: removing

File: tests/unit/CrazyTestCaseTest.php [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        $this->assertParamIsNullable($param);
+    }
+
+    private static function fn3(int $int)
+    {
+    }
+
+    public function testAssertParamIsNotNullable()
+    {
+        $param = self::getFunc('fn3');
```
Comment (Solomonic) [inline tests/unit/CrazyTestCaseTest.php]: why not use `fn1`?

File: tests/unit/CrazyTestCaseTest.php [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        $this->assertParamIsOptional($param);
+    }
+
+    private static function fn5($int)
+    {
+    }
+
+    public function testAssertParamIsNotOptional()
+    {
+        $param = self::getFunc('fn5');
```
Comment (Solomonic) [inline tests/unit/CrazyTestCaseTest.php]: `fn1`?

File: src/TestCaseHelper.php (hunk 13 above)
Comment (dacgray) [reply src/TestCaseHelper.php]: We'll use this in our test classes to help get the ReflectionParameter of the arg we want to test.

We can place it in CrazyTestClass, but I reasoned it falls outside the scope of the class.  It's more of a helper method and so I placed it in it's own class.

File: README.md (hunk 6 above)
Comment (Solomonic) [reply README.md]: The `vcs` package approach is common in the php world and valid in my book to get things up and running esp. with the CI. To use the path-directory with symlink is of course possible too. As long as we are pulling in "production mode" from gemfury or packagist this is a non-issue to me, either way is fine

File: composer.json (hunk 7 above)
Comment (wmathes) [reply composer.json]: I've seen your issue about that. I'd prefer fixing it while at it. Should be a simple switcheroo :P

**Comment by adhocore** [comment]
i dont see any use case of this repo, catching exception is already supported by phpunit
besides the [Throwable Error](http://php.net/Error) in general is something [not to be tested/asserted](https://github.com/sebastianbergmann/phpunit/issues/1778#issuecomment-120942848)

**Comment by dacgray** [comment]
It provides 5 very lazy, but comfortable assertions.

The `assertExeptionThrown` test makes it easy to test for multiple exceptions in one test method.  Yes, PHPUnit forces one test method per exception test because code should break after an exception, and we will need to watch out for this, but this provides a nice comfort/readability advantage.

The param tests are also there mostly for comfort/readability. Strict testing requires you to test implementation - and these tests simply test how you wrote your method, but I think there is a valid use case in PHP world.  

Consider a test on `function fn1(int $i){}` and `function fn2(int $i){}` If it's ambiguous whether $i will be a string/int, we either need to write TypeError exception tests or data validation tests to catch if a programmer switches from int to string or vice versa.  The very lazy alternative is to have type hint tests.

**Comment by adhocore** [comment]
parameter typehint tests are actually not required (exactly what phpunit creator says). 
it may sometimes be required for return type hints- the only case is when we have something like:

```php
function fn($case)
{
// some case: return 1;
// another case: return null;
// yet another: return [];
}

// test will be:
$this->assertInternalType('int', fn(some));
$this->assertNull(fn(another));
$this->assertInternalType('array', fn(yet another));
```

`function fn1(int $i){}` is not ambiguous because it tells `int $i`, but `fn1($i){}` is ambigious.
